src/crimson/net/ProtocolV2.cc

```diff
@@ -0,0 +1,1603 @@
+                         " flags={} cookie={}",
+                         conn,
+                         server_ident.addrs(), server_ident.gid(),
+                         server_ident.global_seq(),
+                         server_ident.supported_features(),
+                         server_ident.required_features(),
+                         server_ident.flags(), server_ident.cookie());
+
+          // is this who we intended to talk to?
+          // be a bit forgiving here, since we may be connecting based on addresses parsed out
+          // of mon_host or something.
+          if (!server_ident.addrs().contains(conn.target_addr)) {
+            logger().warn("{} peer identifies as {}, does not include {}",
+                          conn, server_ident.addrs(), conn.target_addr);
+            abort_in_fault();
+          }
+
+          server_cookie = server_ident.cookie();
+
+          // TODO: change peer_addr to entity_addrvec_t
+          ceph_assert(conn.peer_addr == server_ident.addrs().front());
+          peer_name = entity_name_t(conn.get_peer_type(), server_ident.gid());
+          conn.set_features(server_ident.supported_features() &
+                            conn.policy.features_supported);
+          peer_global_seq = server_ident.global_seq();
+
+          // TODO: lossless policy
+          ceph_assert(server_ident.flags() & CEPH_MSG_CONNECT_LOSSY);
+          conn.policy.lossy = server_ident.flags() & CEPH_MSG_CONNECT_LOSSY;
+          // TODO: backoff = utime_t();
+          logger().debug("{} connect success {}, lossy={}, features={}",
+                         conn, connect_seq, conn.policy.lossy, conn.features);
+
+          return dispatcher.ms_handle_connect(
+              seastar::static_pointer_cast<SocketConnection>(conn.shared_from_this()))
+          .handle_exception([this] (std::exception_ptr eptr) {
+            logger().error("{} ms_handle_connect caust exception: {}", conn, eptr);
+          });
+        }).then([this] {
+          return true;
+        });
+      default: {
+        unexpected_tag(tag, conn, "post_client_connect");
+        // won't be executed
+        return seastar::make_ready_future<bool>(false);
+      }
+    }
+  });
+}
+
+seastar::future<bool> ProtocolV2::client_reconnect()
+{
+  // send_reconnect() logic
+  auto reconnect = ReconnectFrame::Encode(messenger.get_myaddrs(),
+                                          client_cookie,
+                                          server_cookie,
+                                          global_seq,
+                                          connect_seq,
+                                          conn.in_seq);
+  logger().debug("{} reconnect to session: client_cookie={}"
+                 " server_cookie={} gs={} cs={} ms={}",
+                 conn, client_cookie, server_cookie,
+                 global_seq, connect_seq, conn.in_seq);
+  return write_frame(reconnect)
+  .then([this] {
```

Comment:
might want to move `.then([this] {` into the previous line. so the code is more compacted.

Reply:
fixed